### ISSUE TITLE
Flatpak Networking Broken - Ping Not Found

### ISSUE THREAD

**Issue by kii-chan-reloaded** [koreader/koreader#12571]
* KOReader version: 2024.07 (Flatpak)
* Device: Fedora Silverblue / Debian Bookworm

#### Issue

Networking in the Flatpak version does not work from a clean install due to a missing `ping` executable (Underlying cause of #11305)

#### Steps to reproduce

Download the Flatpak from Flathub and run `flatpak run rocks.koreader.KOReader`. The output in the terminal will contain the line `sh: line 1: ping: command not found` (emphasized in the log below)

#### Fix

The KOReader Flatpak is set up with the `filesystem=host` portal by default, so the system's `ping` executable can be found under `/run/host/usr/bin`. As such, the Flatpak's `PATH` environment variable should be updated to be `PATH=/app/bin:/usr/bin:/run/host/usr/bin:/run/host/bin`.

This fix can be done easily by the user via `flatpak override` or Flatseal, but it should be configured this way "out of the box".

#### Log Output

```
 [*] Current time: 09/29/24-23:19:01
 [*] Version: v2024.07

ffi.load: libs/libutf8proc.so.3
ffi.load: blitbuffer
ffi.load: SDL2
ffi.load: SDL2
Starting SDL in /app/koreader/
ffi.load: libs/liblodepng.so
09/29/2024-11:19:01 PM INFO  initializing for device Flatpak 
09/29/2024-11:19:01 PM INFO  framebuffer resolution: {
  h = 1400,
  w = 2174
} --[[table: 0x7f2a9e1ec708]] 
ffi.load: libs/libwrap-mupdf.so
ffi.load: sqlite3
ffi.load: libs/libfreetype.so.6
ffi.load: libs/libharfbuzz.so.0
ffi.load: libs/libzstd.so.1

sh: line 1: ping: command not found

09/29/2024-11:19:01 PM INFO  Loading plugins from directory: plugins 
09/29/2024-11:19:01 PM INFO  Loading plugins from directory: /var/home/______/.var/app/rocks.koreader.KOReader/config/koreader/plugins/ 
```

**Comment by Frenzie** [comment]
> The KOReader Flatpak is set up with the filesystem=host portal by default, so the system's ping executable can be found under /run/host/usr/bin. As such, the Flatpak's PATH environment variable should be updated to be PATH=/app/bin:/usr/bin:/run/host/usr/bin:/run/host/bin.

I'm sorry, but that's nonsensical.

In any case, pinging @proletarius101.

**Comment by Frenzie** [comment]
> use the desktop-agnostic XDG desktop portal

That is very non-universal ftr. ;-) 

> ping with lua libraries

We do that too. I don't know why this one here doesn't; could be worth checking out.

**Comment by kii-chan-reloaded** [comment]
I don't think that changing the code of KOReader is necessary for this issue to be resolved, and to be fair I also don't think that this is a bug in Flatpak either. The only thing that needs to change is the Flatpak's manifest/build command; wherever `--filesystem=host` is being declared during the build process, there should also be `--env=PATH=/app/bin:/usr/bin:/run/host/usr/bin:/run/host/bin`. I had wanted to submit a pull request originally to just knock this out in one go, but I couldn't find its `manifest.json` in this repo or the few other KOReader repos I checked.

**Comment by pazos** [comment]
> Agreed. But using the ping binary to decide whether the network works is unnecessary. I would either

* use the desktop-agnostic XDG desktop portal [org.freedesktop.portal.NetworkMonitor.GetStatus](https://flatpak.github.io/xdg-desktop-portal/docs/doc-org.freedesktop.portal.NetworkMonitor.html#org-freedesktop-portal-networkmonitor-getstatus);
*  or ping with lua libraries

The former isn't possible without a bunch of work.
The later is, IIRC, not possible wihout root (or suid'ing the luajit binary)

**Comment by pazos** [comment]
> I don't think that changing the code of KOReader is necessary for this issue to be resolved, and to be fair I also don't think that this is a bug in Flatpak either. The only thing that needs to change is the Flatpak's manifest/build command; wherever `--filesystem=host` is being declared during the build process, there should also be `--env=PATH=/app/bin:/usr/bin:/run/host/usr/bin:/run/host/bin`. I had wanted to submit a pull request originally to just knock this out in one go, but I couldn't find its `manifest.json` in this repo or the few other KOReader repos I checked.

It is in https://github.com/flathub/rocks.koreader.KOReader

> there should also be `--env=PATH=/app/bin:/usr/bin:/run/host/usr/bin:/run/host/bin`. 

Is this construct already used on other flatpak packages?. Could you share an example of an app that export /run/host/bin as part of its path?

The solution seems easier than the two above. But still needs some thinking,  specially if calling binaries from that path isn't a common procedure on flatpaks, since it can be removed at any point in time.

**Comment by Frenzie** [comment]
> and to be fair I also don't think that this is a bug in Flatpak either

I strongly disagree. A broken system is a bug and the idea that you'd need to take special action to unbreak that which works everywhere it is preposterous. If this is not considered a bug it means I consider Flatpak itself a bug, plain and simple.

**Comment by kii-chan-reloaded** [comment]
> > there should also be `--env=PATH=/app/bin:/usr/bin:/run/host/usr/bin:/run/host/bin`.
> 
> Is this construct already used on other flatpak packages?. Could you share an example of an app that export /run/host/bin as part of its path?
> 
> The solution seems easier than the two above. But still needs some thinking, specially if calling binaries from that path isn't a common procedure on flatpaks, since it can be removed at any point in time.

I've only played around with Flatpak development for personal projects, I don't have any idea what the best practices are with regards to missing BusyBox binaries. I suppose I can look around.

Would it be better to shadow the `PATH` variable for the ping command specifically? What I mean is changing

https://github.com/koreader/koreader/blob/e2d1c55ec56b50c929efba324e8d84a9e337a56d/frontend/device/sdl/device.lua#L395

to be 

```lua
return 0 == os.execute("PATH=$PATH:/run/host/usr/bin:/run/host/bin ping -c1 -w2 " .. default_gw .. " > /dev/null")
```

Nonexistent `PATH` folders don't cause errors, so there would be no real harm in doing it this way.

----

> I strongly disagree. A broken system is a bug and the idea that you'd need to take special action to unbreak that which works everywhere it is preposterous. If this is not considered a bug it means I consider Flatpak itself a bug, plain and simple.

Calling "Flatpak itself" a bug is a bit extreme. Flatpak defaults its sandbox to the minimum possible requirements to make the risk of system/personal damage from a misbehaving or malicious program. Access to networking as a whole is off by default in Flatpak, so it makes some sense that their base images don't include things like `ping`. All the`filesystem=host` declaration does is make a mount point at `/run/host`, anything past that is up to the package maintainer to configure. That's just... how Flatpaks work.

**Comment by NiLuJe** [comment]
> The later is, IIRC, not possible wihout root (or suid'ing the luajit binary)

Oh, right, I'd forgotten about this little bit of brain damage.

c.f., https://github.com/koreader/koreader/blob/e2d1c55ec56b50c929efba324e8d84a9e337a56d/frontend/device/generic/device.lua#L637-L668

**Comment by Frenzie** [comment]
> so it makes some sense that their base images don't include things like ping

The implication is that nothing that is expected to work works, in the genre of cat or wc. If the framing of "things like ping" is accurate that might potentially be slightly different.

**Comment by pazos** [comment]
Rely on the dns resolver instead of pinging it is also impossible because most of the desktop linux will rely on systemd-resolved, which is known to be broken on flatpak.

Anyhow, it seems that system ping sucks less than reinventing the wheel.

> Would it be better to shadow the PATH variable for the ping command specifically?

Not really, but you could try your original suggestion (add to PATH) and see if that, at least, work.

Most network features should work fine now (wikipedia, dicts...) . The issue seem to affect actions that check connectivity before deciding between execution and scheduling for later.

**Comment by pazos** [comment]
Great, thanks

> Personally, I think shadowing PATH for the ping command makes the most sense in this scenario if it would be frowned upon to just add the host's /usr/bin to the entire Flatpak's PATH.

Agree with the intention but it is a bit obscure. Better to make it clear all this nonsense happens just on flatpak.

Will submit something later, when I get access to a real computer

**Comment by pazos** [comment]
Actually, nope.

It is a bug on flatpak to choose not to inherit the host environment when `--filesystem=host`  is defined. At least the PATH should be preserved in that case.

It is also the intended behaviour, as of https://github.com/flatpak/flatpak/issues/4280#issuecomment-843522577, so it is totally fine for us to set the PATH again after `flatpak-run` unsets it. That what we're doing with `--env=LD_LIBRARY_PATH=/app/lib` as `LD_LIBRARY_PATH` is also unset.

Doing the changes in the glue code has some other advantages:

- Makes the fix available as soon as possible, instead of waiting for the next tagged release.
- Makes possible to use common logic to check if unsandboxed apps are available, rendering https://github.com/koreader/koreader/blob/master/frontend/device/sdl/device.lua#L134 obsolete.

Ideally `flatpak` should provide a method to check network connectivity for legacy apps (like they're doing with `--filesystem=xdg-run/gvfs` to access remote mountpoints on non-GIO and legacy apps)

Feel free to open tickets upstream, if you wish.
Since we can workaround here and the `/run/host` thing is documented in https://docs.flatpak.org/en/latest/sandbox-permissions.html we can set our PATH freely, wishing better integration in the future :)


**tl;dr**: @kii-chan-reloaded, feel free to submit the patch suggested in https://github.com/koreader/koreader/issues/12571#issuecomment-2383273108 against https://github.com/flathub/rocks.koreader.KOReader.
